Code/GraphMol/Resonance.cpp

```diff
@@ -0,0 +1,1355 @@
+            break;
+        }
+        s += " failed";
+        throw std::runtime_error(s);
+      }
+    }
+    
+    // fix the number of explicit and implicit Hs in the
+    // resonance structure which has been assembled
+    void fixExplicitImplicitHs(ROMol &mol)
+    {
+      mol.clearComputedProps(false);
+      for (ROMol::AtomIterator ai = mol.beginAtoms();
+           ai != mol.endAtoms(); ++ai) {
+        (*ai)->clearComputedProps();
+        (*ai)->setNumExplicitHs((*ai)->getNumImplicitHs()
+             + (*ai)->getNumExplicitHs());
+        (*ai)->updatePropertyCache();
+      }
+    }
+
+  }
+
+  // object constructor
+  AtomElectrons::AtomElectrons(ConjElectrons *parent, const Atom *a) :
+    d_nb(0),
+    d_tv(static_cast<boost::uint8_t>(a->getTotalDegree())),
+    d_fc(0),
+    d_flags(0),
+    d_atom(a),
+    d_parent(parent)
+  {
+  }
+  
+  // copy constructor
+  AtomElectrons::AtomElectrons(ConjElectrons *parent, const AtomElectrons &ae) :
+    d_nb(ae.d_nb),
+    d_tv(ae.d_tv),
+    d_fc(ae.d_fc),
+    d_flags(ae.d_flags),
+    d_atom(ae.d_atom),
+    d_parent(parent)
+  {
+  }
+  
+  // assign total valence, formal charge and non-bonded electrons
+  // from the original atom
+  void AtomElectrons::initTvNbFcFromAtom()
+  {
+    d_tv = d_atom->getTotalValence();
+    d_fc = d_atom->getFormalCharge();
+    d_nb = oe() - d_tv - d_fc;
+  }
+  
+  boost::uint8_t AtomElectrons::findAllowedBonds(unsigned int bi)
+  {
+    // AtomElectrons::findAllowedBonds returns a 6-bit result
+    // encoded as follows:
+    // +-------------------------------------+
+    // | BIT | MEANING                       |
+    // |   1 | can accept single bond        |
+    // |   2 | needs charge if single bonded |
+    // |   3 | can accept double bond        |
+    // |   4 | needs charge if double bonded |
+    // |   5 | can accept triple bond        |
+    // |   6 | needs charge if triple bonded |
+    // +-------------------------------------+
+    
+    allConjBondsDefinitiveBut(bi);
+    boost::uint8_t res = 0;
+    for (unsigned int i = 0; i < 3; ++i) {
+      res |= (canAddBondWithOrder(bi, i + 1) << (i * 2));
+    }
+    return res;
+  }
+
+  // returns true if any conjugated neighbor is charged (and
+  // has atomic number == atomicNum, if the latter is non-zero)
+  bool AtomElectrons::isNbrCharged(unsigned int bo,
+    unsigned int atomicNum)
+  {
+    bool res = false;
+    const ROMol &mol = d_parent->parent()->mol();
+    ROMol::OEDGE_ITER nbrIdx, endNbrs;
+    boost::tie(nbrIdx, endNbrs) = mol.getAtomBonds(d_atom);
+    for (; !res && (nbrIdx != endNbrs); ++nbrIdx) {
+      const Bond *bondNbr = mol[*nbrIdx].get();
+      unsigned int biNbr = bondNbr->getIdx();
+      if (d_parent->parent()->getBondConjGrpIdx(biNbr) != conjGrpIdx())
+        continue;
+      BondElectrons *beNbr =
+        d_parent->getBondElectronsWithIdx(biNbr);
+      const Atom *atomNbr = ((bondNbr->getBeginAtom() == d_atom)
```

Comment:
Hi Greg,

Thanks for your feedback, I'll make those changes and check the hashing behaviour in Boost.

Cheers,
p.

> On 13 Oct 2015, at 06:25, Greg Landrum <notifications@github.com> wrote:
> 
> In Code/GraphMol/Resonance.cpp:
> 
> > +  // has atomic number == atomicNum, if the latter is non-zero)
> > +  bool AtomElectrons::isNbrCharged(unsigned int bo,
> > +    unsigned int atomicNum)
> > +  {
> > +    bool res = false;
> > +    const ROMol &mol = d_parent->parent()->mol();
> > +    ROMol::OEDGE_ITER nbrIdx, endNbrs;
> > +    boost::tie(nbrIdx, endNbrs) = mol.getAtomBonds(d_atom);
> > +    for (; !res && (nbrIdx != endNbrs); ++nbrIdx) {
> > +      const Bond *bondNbr = mol[*nbrIdx].get();
> > +      unsigned int biNbr = bondNbr->getIdx();
> > +      if (d_parent->parent()->getBondConjGrpIdx(biNbr) != conjGrpIdx())
> > +        continue;
> > +      BondElectrons *beNbr =
> > +        d_parent->getBondElectronsWithIdx(biNbr);
> > +      const Atom *atomNbr = ((bondNbr->getBeginAtom() == d_atom)
> This is more simply done with Bond::getOtherAtom()
> 
> —
> Reply to this email directly or view it on GitHub.
>